src/brod_client.erl

```diff
@@ -664,30 +708,93 @@ do_get_partitions_count(TopicMetadata) ->
     false -> {ok, erlang:length(Partitions)}
   end.
 
+-spec maybe_start_producer(client(), topic(), partition(), {error, any()}) ->
+        ok | {error, any()}.
 maybe_start_producer(Client, Topic, Partition, Error) ->
-  case gen_server:call(Client, {auto_start_producer, Topic}) of
-    {error, topic_not_found} = Error ->
-      Error;
-    false ->
-      Error;
-    {error, {already_started, _}} ->
-      get_partition_worker(Client, ?PRODUCER_KEY(Topic, Partition));
+  case safe_gen_call(Client, {auto_start_producer, Topic}, infinity) of
     ok ->
-      get_partition_worker(Client, ?PRODUCER_KEY(Topic, Partition))
+      get_partition_worker(Client, ?PRODUCER_KEY(Topic, Partition));
+    disabled ->
+      Error;
+    {error, Reason} ->
+      {error, Reason}
+  end.
+
+-spec send_sync(#state{}, kpro_Request(), timer:time()) ->
+        {Result, #state{}} when Result :: {ok, kpro_Response()}
+                                        | {error, any()}.
+send_sync(State, Request, Timeout) ->
+  send_sync(State, Request, Timeout, 0).
+
+send_sync(State, Request, Timeout, RetryCnt) ->
+  RetryCnt >= length(State#state.bootstrap_endpoints) andalso
+    erlang:error(no_reachable_endpoint),
```

Comment:
we have a rotation in endpoint list.
when 'EXIT' is received, or when {error, _} returned in brod_sock:start_link call.
meaning, maybe_restart_metadata_socket will most likely try the next one in the list.

However your comment reminds me that 
we should perhaps react on the error here and move current to the endpoint list rear.